core/src/main/java/com/datastax/oss/driver/api/core/metadata/Metadata.java

```diff
@@ -126,4 +133,51 @@
   default Optional<String> getClusterName() {
     return Optional.empty();
   }
+
+  /**
+   * Generates a new {@link TopologyDiagnostic} for this metadata. See the javadocs of {@link
+   * TopologyDiagnostic} for a detailed explanation on the contents of this diagnostic.
+   *
+   * <p>Note: diagnostics rely on Gossip events received by the driver. But Gossip events are not
+   * 100% reliable, and therefore, the accuracy of reported health statuses should be considered
+   * best-effort only, and are not meant to replace a proper operational surveillance tool.
+   */
+  default TopologyDiagnostic generateTopologyDiagnostic() {
+    return new TopologyDiagnosticGenerator(this).generate();
+  }
+
+  /**
+   * Generates a new {@link TokenRingDiagnostic} for this metadata, analyzing token range
+   * availability for the provided keyspace, consistency level and datacenter. See the javadocs of
+   * {@link TokenRingDiagnostic} for a detailed explanation on the contents of this diagnostic.
+   *
+   * <p>For this method to be able to generate a diagnostic, the following conditions need to be
+   * met:
+   *
+   * <ol>
+   *   <li>Token metadata is enabled;
+   *   <li>The keyspace has a supported replication strategy;
+   *   <li>The datacenter name is non-null (required only for {@linkplain
+   *       ConsistencyLevel#isDcLocal() datacenter-local consistency levels}).
+   * </ol>
+   *
+   * <p>Note: diagnostics rely on Gossip events received by the driver. But Gossip events are not
+   * 100% reliable, and therefore, the accuracy of reported health statuses should be considered
+   * best-effort only, and are not meant to replace a proper operational surveillance tool.
+   *
+   * @param keyspace the {@link KeyspaceMetadata keyspace} to analyze.
+   * @param consistencyLevel the {@link ConsistencyLevel consistency level} to use.
+   * @param datacenter the datacenter name; only required for datacenter-local consistency levels,
+   *     may be null otherwise.
+   * @see DefaultDriverOption#METADATA_TOKEN_MAP_ENABLED
+   * @see #getKeyspace(String)
+   */
+  default Optional<TokenRingDiagnostic> generateTokenRingDiagnostic(
```

Comment:
I wonder if throwing an error would be better than returning `Optional.empty()`.

Reply:
👍 There are a lot of error cases where you just log warnings and return empty optionals/collections. I think it makes it harder to realize that something went wrong, it would be more efficient to surface them directly as exceptions.

Reply:
But... they aren't exception cases.  The empty optional is returned here if something is _disabled_.  How is that worth an exception?

I'd argue the Optional usage is right, at least for that case.  If you have something that's truly worthy of an exception you could do that.  You _might_ want to throw an exception for the unsupported replication strategy case, for instance... I could see that more readily than the "token map is disabled" case.

Reply:
The problem with the empty optional is that it doesn't surface the error to the caller. Your only option is to go sift through the logs to find the message (hoping warn level is enabled). There's no convenient way to do this programmatically either: for example if the diagnostic was requested from a GUI and you'd like to let the user know why it's not showing up, it's going to be a nightmare to extract that message.

I realize that this problem already exists in some parts of the API: for example, it might have been useful to know why `Metadata.getTokenMap()` is absent... But here it seems more obvious because it's an operation that's invoked explicitly, not something the driver does in the background.

PS- before someone brings it up: no, we don't need a Try monad 😉

Reply:
OK I'm changing this to throw exceptions.